pkg/controller/controller.go

```diff
@@ -198,21 +224,76 @@ func (c *Controller) Run(threadiness int, stopCh chan struct{}) error {
 		return fmt.Errorf("failed to wait for caches to sync")
 	}
 
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		select {
+		case <-stopCh:
+			cancel()
+		case <-ctx.Done():
+		}
+	}()
+	if c.leaderElection {
+		return c.runWorkersWithLeaderElection(ctx, threadiness)
+	}
+	c.runWorkers(ctx, threadiness)
+	return nil
+}
+
+func (c *Controller) runWorkersWithLeaderElection(ctx context.Context, threadiness int) error {
+	leaderElectionNamespace := c.leaderElectionNamespace
+	if leaderElectionNamespace == "" {
+		var err error
+		if leaderElectionNamespace, err = getInClusterNamespace(); err != nil {
+			return err
+		}
+	}
+
+	leaderElectionIdentity := string(uuid.NewUUID())
+	leaderElectionLock, err := resourcelock.New(resourcelock.ConfigMapsResourceLock,
+		leaderElectionNamespace,
+		c.leaderElectionID,
+		c.kubeclientset.CoreV1(),
+		c.kubeclientset.CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity: leaderElectionIdentity,
+		})
+	if err != nil {
+		return err
+	}
+	elector, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
+		Lock:          leaderElectionLock,
+		LeaseDuration: 60 * time.Second,
+		RenewDeadline: 15 * time.Second,
+		RetryPeriod:   5 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				c.runWorkers(ctx, threadiness)
+			},
+			OnStoppedLeading: func() {
+				klog.Info("losing leader")
+			},
```

Comment:
Will the workers stop automatically if it's no longer the leader? Where is `stopCh` closed?

Reply:
yes, the stopCh didn't close, but the process will finish. e.g. the elector.Run(ctx) will finish :D